x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/inference/InferenceRunner.java

```diff
@@ -0,0 +1,155 @@
+import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsConfig;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelConfig;
+import org.elasticsearch.xpack.core.ml.inference.results.InferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.inference.InferenceDefinition;
+import org.elasticsearch.xpack.core.ml.utils.MapHelper;
+import org.elasticsearch.xpack.ml.dataframe.DestinationIndex;
+import org.elasticsearch.xpack.ml.dataframe.stats.DataCountsTracker;
+import org.elasticsearch.xpack.ml.dataframe.stats.ProgressTracker;
+import org.elasticsearch.xpack.ml.inference.loadingservice.Model;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelProvider;
+import org.elasticsearch.xpack.ml.utils.persistence.ResultsPersisterService;
+
+import java.io.IOException;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class InferenceRunner {
+
+    private static final Logger LOGGER = LogManager.getLogger(InferenceRunner.class);
+
+    private static final int MAX_PROGRESS_BEFORE_COMPLETION = 98;
+    private static final int RESULTS_BATCH_SIZE = 1000;
+
+    private final Client client;
+    private final TrainedModelProvider modelProvider;
+    private final ResultsPersisterService resultsPersisterService;
+    private final TaskId parentTaskId;
+    private final DataFrameAnalyticsConfig config;
+    private final ProgressTracker progressTracker;
+    private final DataCountsTracker dataCountsTracker;
+    private volatile boolean isCancelled;
+
+    public InferenceRunner(Client client, TrainedModelProvider modelProvider, ResultsPersisterService resultsPersisterService,
+                           TaskId parentTaskId, DataFrameAnalyticsConfig config, ProgressTracker progressTracker,
+                           DataCountsTracker dataCountsTracker) {
+        this.client = Objects.requireNonNull(client);
+        this.modelProvider = Objects.requireNonNull(modelProvider);
+        this.resultsPersisterService = Objects.requireNonNull(resultsPersisterService);
+        this.parentTaskId = Objects.requireNonNull(parentTaskId);
+        this.config = Objects.requireNonNull(config);
+        this.progressTracker = Objects.requireNonNull(progressTracker);
+        this.dataCountsTracker = Objects.requireNonNull(dataCountsTracker);
+    }
+
+    public void cancel() {
+        isCancelled = true;
+    }
+
+    public void run(TrainedModelConfig trainedModelConfig, InferenceConfig inferenceConfig) {
+        if (isCancelled) {
+            return;
+        }
+
+        LOGGER.info("[{}] Running inference on model [{}]", config.getId(), trainedModelConfig.getModelId());
+
+        try {
+            InferenceDefinition inferenceDefinition = modelProvider.inflateModelForInference(trainedModelConfig);
```

Comment:
Currently, we keep a reference of the latest model read by the job and then we directly inflate it here. The intention is to change this and use `ModelLoadingService` instead.